### PR TITLE
ICU-584 Add support for own DM

### PR DESCRIPTION
**Pull request by saturninoabril**
#### Summary
Add support for own DM which is the same as implemented in webapp
- include current user to the list of users when adding more direct channels
- clicking own user directly creates/opens own DM channel
- when another user is initially selected when creating direct channel, current user is immediately removed from the list of users
- list of users behave as expected when removing, selecting or searching for users

#### Ticket Link
Jira ticket: [ICU-584](https://mattermost.atlassian.net/browse/ICU-584)

#### Checklist
- [x] Has UI changes
- [x] Includes text changes and localization file updates

#### Device Information
This PR was tested on: [iPhone 6 simulator, IOS 11.2] - Update: also tested to Android Emulator Nexus 5x API 23

#### Screenshots
screen shot 2018-01-17 at 4 44 06 am
screen shot 2018-01-17 at 4 44 21 am

### CODE REVIEW

**Comment by enahum** [comment]
@stephenkiers when that label is applied means that the person who added the label wants this PR to be tested in Android and iOS unless explicitly asked for a platform.

that means that for now developers should not only review the code but the functionality as well, by checking out the PR and running it. At least it was like that until now, but with the work done by @mkraft we should be building the PR so it can be tested on the devices, not sure about the process here as we haven't done that even once

File: app/components/channel_drawer/channels_list/channel_item/channel_item.js [hunk 1]
```diff
@@ -81,6 +85,16 @@ export default class ChannelItem extends PureComponent {
             type
         } = this.props;
 
+        const {intl} = this.context;
+
+        let channelDisplayName = displayName;
+        if (currentUserId === channelTeammateId) {
+            channelDisplayName = intl.formatMessage({
+                id: 'mobile.channel.you',
```
Comment (enahum) [inline app/components/channel_drawer/channels_list/channel_item/channel_item.js]: I think here you could just use `channel_header.directchannel.you` as it already exists and is the same message

File: app/components/channel_drawer/channels_list/channel_item/index.js [hunk 2]
```diff
@@ -18,8 +19,15 @@ function makeMapStateToProps() {
             member = getMyChannelMember(state, ownProps.channelId);
         }
 
+        let teammate = null;
+        if (channel.teammate_id) {
+            teammate = getUser(state, channel.teammate_id);
```
Comment (enahum) [inline app/components/channel_drawer/channels_list/channel_item/index.js]: Why are we getting the user if we only need the id which we already have?

more on that don't you think instead of passing the `currentUserId` and `channelTeammateId` as props you can pass something like `isMyUser` that is equals to `currentUserId === channel.teammate_id`?

Comment (enahum) [reply app/components/channel_drawer/channels_list/channel_item/index.js]: one other thing I'm seeing **channel.teammate_id** and we definitely shouldn't be using that as teammate_id is not a property of the channel and is being injected by a selector somewhere that we should remove.

Perhaps check if the channel.type is a DM and then get the teammate_id from it.

File: app/components/channel_intro/index.js [hunk 3]
```diff
@@ -37,28 +25,22 @@ function makeMapStateToProps() {
...
-
-            return [otherUser];
-        }
-    );
-
     return function mapStateToProps(state, ownProps) {
         const currentChannel = getChannel(state, {id: ownProps.channelId}) || {};
         const {status: getPostsRequestStatus} = state.requests.posts.getPosts;
 
+        const teammate = [];
```
Comment (enahum) [inline app/components/channel_intro/index.js]: I really want to revert the changes here as we should not use `channel.teammate_id` I've explained the reason above

File: app/components/custom_list/user_list_row/user_list_row.js [hunk 4]
```diff
@@ -17,36 +18,62 @@ import {displayUsername} from 'mattermost-redux/utils/user_utils';
...
             <CustomListRow
-                id={this.props.id}
-                theme={this.props.theme}
-                onPress={this.props.onPress ? this.onPress : null}
-                enabled={this.props.enabled}
-                selectable={this.props.selectable}
-                selected={this.props.selected}
+                id={id}
+                theme={theme}
+                onPress={onPress ? this.onPress : null}
```
Comment (enahum) [inline app/components/custom_list/user_list_row/user_list_row.js]: I think you should pass `this.onPress` and in the function validate that the prop exists

File: app/screens/more_dms/more_dms.js [hunk 5]
```diff
@@ -208,42 +231,81 @@ class MoreDirectMessages extends PureComponent {
...
 
-            return {
-                selectedIds,
-                selectedCount: Object.keys(selectedIds).length
-            };
-        });
+                let newProfiles = profiles;
+                if (wasSelected) {
+                    Reflect.deleteProperty(newSelectedIds, id);
+                    if (selectedCount === 1) {
```
Comment (enahum) [inline app/screens/more_dms/more_dms.js]: i don't quite understand this

File: app/components/channel_drawer/channels_list/channel_item/index.js (hunk 2 above)
Comment (saturninoabril) [reply app/components/channel_drawer/channels_list/channel_item/index.js]: Agree on your points and teammate is not necessary at this stage. But this will come handy when we implement the "DM channel with deactivated users" feature similar to webapp https://mattermost.atlassian.net/browse/PLT-7917.
I'll update the change as you commented above. Thanks!

File: app/screens/more_dms/more_dms.js (hunk 5 above)
Comment (saturninoabril) [reply app/screens/more_dms/more_dms.js]: This section handles selected user from 1 to 0. This condition resets the slice of profiles including the user's self.

File: app/components/channel_intro/index.js [hunk 6]
```diff
@@ -24,7 +24,7 @@ function makeMapStateToProps() {
                 return '';
             }
 
-            return channel.name.split('__').find((m) => m !== currentUserId);
+            return channel.name.split('__').find((m) => m !== currentUserId) || currentUserId;
```
Comment (enahum) [inline app/components/channel_intro/index.js]: Why are you assuming that the teammate id here is currentUserId ?

Comment (saturninoabril) [reply app/components/channel_intro/index.js]: For own DM channel's name, it is like `8xauxmth4tnb3jrky4thfkspay__8xauxmth4tnb3jrky4thfkspay` where `8xauxmth4tnb3jrky4thfkspay` is self's user ID. It will return no match so the currentUserId is the the teammate id.
I think it's better to rely on `channel.teammate_id` when channel type is DM, then get the user profile out of it, than to have this function `getChannelMembersForDm` and `getOtherUserIdForDm` like what I did on first commit.

Comment (enahum) [reply app/components/channel_intro/index.js]: actually we cannot rely on `channel.teammate_id` as we should be removing that from the getChannel selector as is modifying the object and creating a new object every time is being called so the selector is actually not memoizing anything.

That said, I think this is ok
